libs/SalesforceSDK/src/com/salesforce/androidsdk/ui/LoginViewModel.kt

```diff
@@ -0,0 +1,287 @@
+import com.salesforce.androidsdk.auth.HttpAccess
+import com.salesforce.androidsdk.auth.OAuth2
+import com.salesforce.androidsdk.auth.OAuth2.TokenEndpointResponse
+import com.salesforce.androidsdk.auth.OAuth2.exchangeCode
+import com.salesforce.androidsdk.auth.defaultBuildAccountName
+import com.salesforce.androidsdk.config.BootConfig
+import com.salesforce.androidsdk.security.SalesforceKeyGenerator.getRandom128ByteKey
+import com.salesforce.androidsdk.security.SalesforceKeyGenerator.getSHA256Hash
+import com.salesforce.androidsdk.util.SalesforceSDKLogger.e
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers.IO
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
+import java.net.URI
+
+open class LoginViewModel(val bootConfig: BootConfig): ViewModel() {
+
+    // LiveData
+    val selectedServer = MediatorLiveData<String>()
+    val loginUrl = MediatorLiveData<String>()
+    internal var dynamicBackgroundColor = mutableStateOf(Color.White)
+    internal var dynamicHeaderTextColor = derivedStateOf { if (dynamicBackgroundColor.value.luminance() > 0.5) Color.Black else Color.White }
+    internal var showServerPicker = mutableStateOf(false)
+    internal val defaultTitleText: String
+        get() = if (loginUrl.value == "about:blank") "" else selectedServer.value ?: ""
+
+    // Public Overrideable LiveData
+    open var showTopBar = true
+    open var topBarColor: Color? = null
+    open var titleText: String? = null
+    open var titleComposable: (@Composable () -> Unit)? = null
+    open var loading = mutableStateOf(false)
+
+    // Additional Auth Values
+    protected open var clientId: String = bootConfig.remoteAccessConsumerKey
+    protected open val authorizationDisplayType = SalesforceSDKManager.getInstance().appContext.getString(oauth_display_type)
+
+    /**
+     * Setting this option to true will enable a mode where only a custom tab will be shown.  The first server will be
+     * launched in a custom tab immediately and the user will not be able to switch servers.  The LoginActivity is
+     * ended if the user attempts to back out of the custom tab.
+     *
+     * Since browser based authentication requires PKCE (and therefore the Web Server flow) the User Agent flow
+     * cannot be used while in this mode.
+     */
+    open val singleServerCustomTabActivity = false
+
+    // LoginOptions values
+    var jwt: String? = null
+    var additionalParameters = hashMapOf<String, String>()
+
+
+    val shouldShowBackButton = with(SalesforceSDKManager.getInstance()) {
+        !(userAccountManager.authenticatedUsers.isNullOrEmpty() || biometricAuthenticationManager?.locked ?: false)
+    }
+    val shouldShowBiometricPromptButton = with(SalesforceSDKManager.getInstance().biometricAuthenticationManager) {
+        this?.let { locked && hasBiometricOptedIn() } ?: false
+    }
+    // todo: add shouldShowIDPButton
```

Comment:
I left several TODOs in on purpose, all related to UI code.  I am not sure which of the UI stories I will be doing vs others so it seemed better to just leave them in place so the intent was not lost.